types/massive/index.d.ts

```diff
@@ -42,194 +49,338 @@ declare namespace massive {
+        /** A Database. */
+        db: Database;
+        /** The table or view's name. */
+        name: string;
+        /** The name of the schema owning the table. */
+        schema: string;
+        /** The table's primary key. */
+        pk: string;
+    }
+
+    class Writable extends Readable {
+        /** A database table or other writable object */
+        constructor(spec: WritableSpecification);
+
+        /** Delete a record or records. */
+        destroy(
+            criteria: AnyObject,
+            options?: ResultProcessingOptions
+        ): Promise<any[]>;
 
         /**
-         * Find a document by searching in the body.
-         *
-         * @param criteria A criteria object or primary key value.
+         * Attempts to assemble primary key criteria for a record object representing a row in this table.
+         * The criteria must include the full primary key, and must not invoke any operations.
          */
-        findDoc(criteria?: object | UUID, options?: GenericQueryOptions): Promise<any>;
+        getPkCriteria(record: AnyObject): AnyObject;
+
+        /** Insert a record or records into the table. */
+        insert(
+            data: AnyObject,
+            options?: PersistenceInsertOptions & ResultProcessingOptions
+        ): Promise<AnyObject>;
+        insert(
+            data: AnyObject[],
+            options?: PersistenceInsertOptions & ResultProcessingOptions
+        ): Promise<AnyObject[]>;
 
         /**
-         * Return a single record.
-         *
-         * @param criteria A criteria object or primary key value.
+         * Saves an object.
+         * If the object does not include a value for the table's primary key, this will emit an INSERT to create a new record.
+         * if it does contain the primary key it will emit an UPDATE for the existing record.
+         * Either way, the newest available version of the record will be returned.
+         * This is not a true Postgres upsert! If you need the behavior of ON CONFLICT DO UPDATE, you'll need to use db.query or create an SQL script file.
          */
-        findOne(criteria?: object | UUID, options?: GenericQueryOptions): Promise<any>;
+        save(
+            record: AnyObject,
+            options?: PersistenceInsertOptions &
+                PersistenceUpdateDocOptions &
+                ResultProcessingOptions
+        ): Promise<AnyObject>;
+
+        /** Save a document to the database. This function will create or replace the entire document body. */
+        saveDoc(doc: AnyObject): Promise<AnyObject>;
 
         /**
-         * Determine whether criteria represent a search by primary key.
-         * If a number or uuid are passed, it is assumed to be a primary key value; if an object, it must have only one key, which must specify the primary key column.
+         * Update a record. May be invoked with a complete record (including primary key), or with a criteria object and a map of fields to new values.
```

Comment:
This comment is out of date. `Writable.update` only accepts `criteria` (object or string/numeric pk) and `changes` so I think you just need the one definition where `criteria` is `AnyObject` below.

Reply:
The JSDoc is correct here, it's context-dependent.